### ISSUE TITLE
Lint/RedundantTypeConversion: false positives on Hash#to_h when a block is given

### ISSUE THREAD

**Issue by akimd** [rubocop/rubocop#13839]
## Expected behavior

```ruby
{a: 1, b: 2}.to_h { |k, v| [k.to_s, v.to_s] }
```
```
$ bundler exec rubocop rubocop/to-h.rb
Inspecting 1 file
E

Offenses:

rubocop/to-h.rb:1:1: C: Naming/FileName: The name of this source file (to-h.rb) should use snake_case.
rubocop/to-h.rb:1:1: C: [Correctable] Style/FrozenStringLiteralComment: Missing frozen string literal comment.
{a: 1, b: 2}.to_h { |k, v| [k.to_s, v.to_s] }
^
rubocop/to-h.rb:1:14: E: [Correctable] Lint/RedundantTypeConversion: Redundant to_h detected.
{a: 1, b: 2}.to_h { |k, v| [k.to_s, v.to_s] }
             ^^^^

1 file inspected, 3 offenses detected, 2 offenses autocorrectable
```

This call to `to_h` is not redundant, it should not complain.

## Actual behavior

It does complain.

## Steps to reproduce the problem

See above.

## RuboCop version

```
$ bundler exec rubocop -V
1.72.0 (using Parser 3.3.7.1, rubocop-ast 1.38.0, analyzing as Ruby 3.3, running on ruby 3.3.7) [arm64-darwin24]
  - rubocop-minitest 0.36.0
  - rubocop-performance 1.23.1
```

Cheers!

**Comment by vlad-pisanov** [comment]
Same issue with sets too by the way:

```ruby
Set[1, 2, 3].to_set(&:odd?)
```

`
Lint/RedundantTypeConversion: Redundant to_set detected.
`

**Comment by bbatsov** [comment]
I've always disliked such transformations, so perhaps the bug is a feature in this case. 😅
